File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.js


// 현재 진행된 거리가 목표지점보다 멀어지면 종료
function distance(x,y,r,c) {
    return Math.abs(r-x) + Math.abs(c-y)
}

function solution(n, m, x, y, r, c, k) {
    function dfs(i,j,s,cnt) {
        if(k < cnt + distance(i,j,r,c)) {
            return;
        }
        
        if(i===r && j === c && cnt === k) {
          answer = s
          return
        }
        
        for(let v = 0; v < move.length; v++) {
            let ii = move[v][0] + i
            let jj = move[v][1] + j
            
            if((1 <= ii && ii <= n) && (1 <= jj && jj <= m) && s < answer) {
                dfs(ii,jj, s + str[v], cnt+1)
            }
        }
    }
    
    var answer = 'z';
    const move = [[1,0],[0,-1],[0,1],[-1,0]]
    const str = ['d','l','r','u']
    const dist = Math.abs(r-x) + Math.abs(c-y)
    if(dist > k || (k-dist) % 2 === 1) return 'impossible'
    
    dfs(x,y,'',0)

    return answer
}

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.py
from collections import deque

def distance(x, y, r, c):
    return abs(x - r) + abs(y - c)

def solution(n, m, x, y, r, c, k):
    answer = ''
    move = [(1, 0), (0, -1), (0, 1), (-1, 0)]
    move_directions = ['d', 'l', 'r', 'u']
    q = deque([(x, y, 0, '')])
    
    dist = distance(x, y, r, c)
    
    if k < dist or (k - dist) % 2 == 1:
        return 'impossible'
    
    while q:
        i, j, cnt, s = q.popleft()
        if i == r and j == c and cnt == k:
            return s
        if i == r and j == c and (k - cnt) % 2:
            return 'impossible'
        for idx, v in enumerate(move):
            ii = i + v[0]
            jj = j + v[1]
            if 1 <= ii <= n and 1 <= jj <= m:
                if k < cnt + distance(ii,jj,r,c) + 1:
                    continue
                q.append((ii, jj, cnt + 1, s + move_directions[idx]))
                break
                    
    return 'impossible'
